backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/types/predicate/impl/ElasticsearchTermsPredicate.java

```diff
@@ -0,0 +1,177 @@
+import org.hibernate.search.backend.elasticsearch.search.impl.AbstractElasticsearchCodecAwareSearchValueFieldQueryElementFactory;
+import org.hibernate.search.backend.elasticsearch.search.impl.ElasticsearchSearchContext;
+import org.hibernate.search.backend.elasticsearch.search.impl.ElasticsearchSearchValueFieldContext;
+import org.hibernate.search.backend.elasticsearch.search.predicate.impl.AbstractElasticsearchSingleFieldPredicate;
+import org.hibernate.search.backend.elasticsearch.search.predicate.impl.PredicateRequestContext;
+import org.hibernate.search.backend.elasticsearch.types.codec.impl.ElasticsearchFieldCodec;
+import org.hibernate.search.engine.backend.types.converter.spi.DslConverter;
+import org.hibernate.search.engine.reporting.spi.EventContexts;
+import org.hibernate.search.engine.search.common.ValueConvert;
+import org.hibernate.search.engine.search.predicate.SearchPredicate;
+import org.hibernate.search.engine.search.predicate.spi.TermsPredicateBuilder;
+import org.hibernate.search.util.common.logging.impl.LoggerFactory;
+
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+
+public class ElasticsearchTermsPredicate extends AbstractElasticsearchSingleFieldPredicate {
+
+	private static final Log log = LoggerFactory.make( Log.class, MethodHandles.lookup() );
+
+	private static final JsonObjectAccessor TERMS_ACCESSOR = JsonAccessor.root().property( "terms" ).asObject();
+	private static final JsonObjectAccessor TERM_ACCESSOR = JsonAccessor.root().property( "term" ).asObject();
+
+	// for boolean query:
+	private static final JsonObjectAccessor BOOL_ACCESSOR = JsonAccessor.root().property( "bool" ).asObject();
+	private static final JsonArrayAccessor MUST_ACCESSOR = JsonAccessor.root().property( "must" ).asArray();
+
+	private static final JsonAccessor<JsonElement> VALUE_ACCESSOR = JsonAccessor.root().property( "value" );
+
+	private final JsonElement term;
+	private final JsonElement[] terms;
+	private boolean allMatch;
+
+	public ElasticsearchTermsPredicate(Builder builder) {
+		super( builder );
+		this.term = builder.term;
+		this.terms = builder.terms;
+		this.allMatch = builder.allMatch;
+	}
+
+	@Override
+	protected JsonObject doToJsonQuery(PredicateRequestContext context, JsonObject outerObject,
+			JsonObject innerObject) {
+
+		// single term
+		if ( term != null ) {
+			VALUE_ACCESSOR.set( innerObject, term );
+			JsonObject middleObject = new JsonObject();
+			middleObject.add( absoluteFieldPath, innerObject );
+			TERM_ACCESSOR.set( outerObject, middleObject );
+			return outerObject;
+		}
+
+		// multiple terms in OR
+		if ( !allMatch ) {
+			JsonArray jsonArray = new JsonArray( terms.length );
+			for ( JsonElement element : terms ) {
+				jsonArray.add( element );
+			}
+
+			innerObject.add( absoluteFieldPath, jsonArray );
+			TERMS_ACCESSOR.set( outerObject, innerObject );
+			return outerObject;
+		}
```

Comment:
You can merge those two blocks by leveraging `VALUE_ACCESSOR.add( innerObject, term )`. It will create an array automatically if there already is a term.

Reply:
If we merge the two blocks, we would use the `TERMS_ACCESSOR` for both the single and the multiple arguments case. I think that if there exists a term predicate for the single argument it is better to keep using it for this case.